### PR TITLE
[cpp]A numerical improvement. Refine tests.

### PR DESCRIPTION
**Pull request by hungpham2511**
Changes in this PRs:
- Add randomized tests to compare Python and C++ API
- A minor fix to improve numerical stability.

Checklists:
- [x] Update CHANGELOG.md with a single line describing this PR

In progress: I will update this PR to run the integration test on CI.

### CODE REVIEW

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -18,8 +18,13 @@ option(OPT_MSGPACK "Serialization using msgpack " OFF)
 # Bindings
 option(PYTHON_BINDINGS "Build bindings for Python" ON)
 set(PYTHON_VERSION 3.7 CACHE STRING "Build bindings for Python version")
+set(BUILD_WITH_PINOCCHIO_PYTHON false CACHE BOOL "Force compile with Pinocchio bindings")
+# This option is needed because sometime the flag
+# PINOCCHIO_WITH_PYTHON_INTERFACE is not properly set depsite a python
+# binding is installed.
```
Comment (jmirabel) [inline cpp/CMakeLists.txt:24]: This comes from robotpkg installation and is a bug from Pinocchio. I will report it.

File: tests/integration_tests/test_with_pinocchio.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+    l = np.array([limit,] * robot.model.nv)
+    lja = LinearJointAcceleration(-l, l)
+    lja.discretizationType = Interpolation
+    return lja
+
+
+def generate_random_trajectory(robot, npts, maxd, randskip=0):
+    # Use to skip configuration. Have no idea how to set random seed
+    # in pinocchio.
```
Comment (jmirabel) [inline tests/integration_tests/test_with_pinocchio.py:74]: `pinocchio.seed(int)`

**Comment by jmirabel** [comment]
> Add randomized tests to compare Python and C++ API

Thanks. I was unsure how to put this test in the Python unit test.

File: docs/source/notes.rst [hunk 3]
```diff
@@ -25,10 +25,14 @@ that is if it takes 5ms to solve an instance with 100 gridpoints, then
 most likely `toppra` will take 10ms to solve another instance which
 has 200 gridpoints.
 
-By default, `toppra` select 100 gridpoints.
-
 As a general rule of thumb, the number of gridpoints should be at
-least twice the number of waypoints in the given path.
+least a few times the number of waypoints in the given path. This is
+not a hard rule, depending on whether the waypoints naturally form a
```
Comment (hungpham2511) [inline docs/source/notes.rst:30]: @jmirabel Please note this point. 

In the integration test, I have reduced the number of waypoints from 50 to 5.

**Comment by jmirabel** [comment]
I see that CI runs on xenial. On xenial, robotpkg provides Python 3.5 stuff, but not Python 3.6. So CI should install robotpkg-py35-pinocchio instead.

**Comment by jmirabel** [comment]
The available versions of py??-pinocchio depending on the distribution can be found here: http://robotpkg.openrobots.org/rbulk/robotpkg/math/py-pinocchio/index.html

**Comment by hungpham2511** [comment]
> The available versions of py??-pinocchio depending on the distribution can be found here: http://robotpkg.openrobots.org/rbulk/robotpkg/math/py-pinocchio/index.html

Thanks. I am skipping the pinocchio tests for now, as on the base CI image is running 16.04 and there are only 2.7 and 3.5, which are incompatible with the python codebase.

**Comment by hungpham2511** [comment]
Look like we will need to upgrade the CI image to bionic and test with 3.6 and 2.7.

**Comment by jmirabel** [comment]
Another option is to install Pinocchio from conda. I never tried myself.
https://github.com/conda-forge/pinocchio-feedstock

**Pull request merged**